fuse_constraints/test/test_marginal_constraint.cpp

```diff
@@ -207,11 +207,11 @@ TEST(MarginalConstraint, LocalParameterization)
   fuse_core::Vector1d expected_residuals;
   expected_residuals << 10.581088914;  // 5.0 * 0.15  +  6.0 * -0.2  +  7.0 * 0.433012702   +   8.0
   fuse_core::MatrixXd expected_jacobian1(1, 4);
-  expected_jacobian1 << -13.29593, 3.86083, 9.164586, 12.818822;
+  expected_jacobian1 << 13.29593, -3.86083, -9.164586, -12.818822;
```

Comment:
Why would this test result change?

Reply:
Scratch that, missed this bullet point in your PR description.

Reply:
Yep, it's explained in the description. :smiley:

Reply:
I haven't done a ton of digging into the Ceres Manifold class, but I suspect Ceres implements the Minus() function differently than I do.

https://github.com/ceres-solver/ceres-solver/blob/2.1.0/include/ceres/autodiff_manifold.h#L166-L170
```
  bool Minus(const double* y,
             const double* x,
             double* y_minus_x) const override
```
Based on the parameter names, I assume this results in `y - x` for a linear system.

Whereas the Minus function in fuse is written such that:
https://github.com/locusrobotics/fuse/blob/devel/fuse_core/include/fuse_core/local_parameterization.h#L62-L79
```
if Plus(x1, delta) -> x2, then Minus(x1, x2) -> delta
```
which using Ceres variable names would be
```
  bool Minus(const double* x,
             const double* y,
             double* y_minus_x) const override
```
The reason, such that there is one, for this variable order is that the general form of "subtraction" would be:
X^-1 * Y. For example, when doing frame transformations you would have T1^1 * T2 to compute "T2 - T1".


That would explain why I'm taking the derivative with respect to the second term and not the first. I'm not sure if it explains the sign change or not.

Reply:
That could also explain the difference between `fuse` and `ceres` local parameterization/manifold implementation of `Minus` and its jacobian.

I've just added a couple of new test cases to the local parameterization unit test: https://github.com/locusrobotics/fuse/pull/341/commits/9d2d148da3cb8e3c4c1f603145ff2a656601d474

I'm basically testing the conditions mentioned in https://github.com/locusrobotics/fuse/blob/c8ab2f6e971a21ecf74e9debeef0d74d1c24655d/fuse_core/include/fuse_core/local_parameterization.h#L68-L69, that is:
* `Minus(x, x) -> 0`
* `if Plus(x1, delta) -> x2, then Minus(x1, x2) -> delta`

The tests are passing with `gcc12` + `Ceres 2.1.0` and should also pass on older versions.

I think with this MR I actually changed your implementation to match the one in `ceres::Manifold`. :laughing: 

Is that ok, or do you think that shouldn't be changed?

Reply:
In the grand scheme of things, I want the fuse version to match the Ceres version. It keeps things much easier.
My worry here is about backwards compatibility. Updating the two local parameterizations contained within the fuse repo is easy. But if anyone has created their own local parameterization classes, then strange things are going to happen.

What would be great is if we can make a translation layer that warns the old methods are dead, and that the new methods must be implemented. But I'm not sure how possible that will be without digging into this MR and the Ceres 2.1 changes in more detail.

Reply:
Actually, I just realized I didn't really change the `Minus` implementation of `Orientation2DLocalParameterization` and `Orientation3DLocalParameterization`.

In the corresponding unit test for those local parameterization we also have functors for the **minus** implementations, but they're also the same as the ones in the classes. So I didn't really change that.

I only changed the analytic and autodiff jacobians, so they match the ones in `ceres::Manifold`.

Reply:
@svwilliams I'm going to be working on refactoring this a bit so that it can be backwards compatible with ceres. My thoughts are to implement a fuse wrapper around the new manifold, and adding deprecated attributes to all localparameterization stuff.

The general logic would be: 
1. if ceres < 2.1 -> use localParameterization, manifold not compiled
2. if ceres > 2.1 -> use both manifold and localParameterizations where available (favour manifold), update internal fuse variables to use manifold
3. if ceres > 2.2 -> use manfiold, localParameterization not compiled

Would this work to get a change like this upstream?

Reply:
> My thoughts are to implement a fuse wrapper

Can you expand on that a bit?

The fuse `LocalParameterization` class is a wrapper around the Ceres Solver `LocalParameterization` class.
```
class LocalParameterization : public ceres::LocalParameterization
```

Are you proposing doing something similar with the manifolds?
```
class Manifold : public ceres::Manifold
```
So that fuse would contain a wrapper class for both `ceres::LocalParameterization` and `ceres::Manifold`?

And then every derived `fuse_core::Variable` class would need to be updated to use the `fuse::Manifold` instead of the `fuse::LocalParameterization` based on the detected version of Ceres Solver?
https://github.com/locusrobotics/fuse/blob/devel/fuse_core/include/fuse_core/variable.h#L315

Or did I misunderstand the proposal completely?

Reply:
Yes thats spot on, then i'd add deprecated flags to whatever needs to be deprecated based on the ceres version.

Reply:
I worried this will create a required API change for any external packages that exist out in the world. And further, that the API change is dependent not on the version of Fuse or ROS, but on the version of one of the Fuse dependencies. And that feels odd.

I haven't worked out the details of the following idea. It might not be easy or even possible. But what if we:
* Create the Fuse wrapper class for the Ceres Manifold class: `class Manifold : public ceres::Manifold`
* Change the Fuse LocalParameterization to derive from the Fuse Manifold class: `class LocalParameterization : public fuse_core::Manifold`
* Modify the implementation of the Fuse LocalParameterization class such that the public API remains the same, but internally it mutates methods into the format expected by the Fuse Manifold class. This is the bit that I don't know if it is feasible.
* Modify the Fuse Variable base class to accept a Fuse Manifold pointer instead of a Fuse LocalParameterization. Since the Fuse LocalParameterization class is now derived from the Fuse Manifold, all existing code should still work.

All of those changes would need to be conditional on the Ceres Solver version >= 2.1. But that keeps the scope of conditional compilation down to two classes (LocalParameterization and Manifold), which seems easier than modifying every concrete Variable implementation.

Does that sound workable at all? Do you see any holes in that plan?

Reply:
I'm thinking that would work but maybe with this slight tweak:
1. Implement a fuse Manifold (extend ceres manifold) as specified, and add the required functions so that it's API is the same as ceres localParameterization
2. fuse::LocalParameterization either extends the fuse::Manifold (if ceres >= 2.1) or it extends ceres::LocalParameterization (if ceres < 2.1) and would be implementing the exact same functions in either case

Then no change is required for the Fuse Variable. I'll try to implement something like this quickly and get your thoughts on it so its more clear

Reply:
> fuse::LocalParameterization either extends the fuse::Manifold (if ceres >= 2.1) or it extends ceres::LocalParameterization (if ceres < 2.1)

Agree completely.

> Implement a fuse Manifold (extend ceres manifold) as specified, and add the required functions so that it's API is the same as ceres localParameterization

I'm not sure the `[fuse::Manifold] API is the same as ceres localParameterization` is necessarily required. I _think_ the main difference is the Ceres Manifold class adds a `Minus()` implementation, something that the Fuse LocalParameterization class has always had. The issue is that I defined the Fuse `Minus()` function slightly differently than the Ceres Manifold `Minus()` function. And it might be nice to sync up the Fuse Manifold API with the Ceres Manifold API as much as possible to avoid confusion. But that means the Fuse LocalParameterization would need to translate the Fuse LocalParameterization function into a Ceres Manifold function.

Regardless, starting a quick implementation sounds great. Then we can start having discussions about concrete concerns rather than theory and supposition. Thanks for this.

Reply:
This is how ceres defines the minus function for manifold:

  // Given two points on the manifold, Minus computes the change to x in the
  // tangent space at x, that will take it to y.

So basically finding delta such that: x + delta = y, which is exactly how it's defined in fuse:

https://github.com/locusrobotics/fuse/blob/devel/fuse_core/include/fuse_core/local_parameterization.h#L68-L69 

SO I don't think any conversions need to be done, but maybe just an updating on the documentation?

Reply:
Just so I can get it straight: The main difference is just the ordering of the two variables in the minus function?

i.e. ceres has Minus as:
`minus(x2, x1) -> x2_minus_x1`

and fuse has it as:
`minus(x1, x2) -> x2_minus_x1`

I don't think it'll be possible to automatically convert between the two since the function signature is the exact same. But I could match the convention of ceres Manifold for the fuse Manifold so that users know for the future. And I don't think any existing code outside of fuse would break since as long as the jacobian for minus is implemented correctly for the operation it shouldn't matter?

Reply:
I believe the input parameters are reversed the Fuse `Minus()` implementation versus the Ceres `Minus()` implementation.

In Fuse:
https://github.com/locusrobotics/fuse/blob/devel/fuse_core/include/fuse_core/local_parameterization.h#L68-L69

```
   *  - Minus(x, x) -> 0
   *  - if Plus(x1, delta) -> x2, then Minus(x1, x2) -> delta
```
Or, to write that a slightly different way
```
if Plus(x, delta) -> x_plus_delta, then Minus(x, x_plus_delta) -> delta
```

Whereas in Ceres:
https://github.com/ceres-solver/ceres-solver/blob/master/include/ceres/manifold.h#L77-L78

```
// 1. x_plus_delta = Plus(x, delta)
// 2. delta = Minus(x_plus_delta, x)
```

Reply:
> so this solution would also include lots of conditional compilations.

I believe it will only require conditional compilation of the Manifold, ManifoldAdapter, LocalParameterization, and Variable classes. And all of that is contained inside of the `fuse_core` package. So at least all of the ugliness would be contained.

> Since the minus function would be implemented by the user,and the jacobian along with it, even if there were to mix up the inputs, but their jacobian also mixed up the inputs, the result would be the same (i think).

If you can prove that is true, then I'll happily accept a simpler alternative.

Reply:
There will be some ugliness in the marginal cost function, having to store both Manifolds and LocalParameterizations

Reply:
@svwilliams is there a clang format file I can use to format to this repo's style?

Reply:
I wrote most of the fuse code before I had clang-format configured. As such, any config file I give you will result in unintended format changes.

But here is my personal, over-engineered clang format file...renamed to make github happy. If that helps you any.
[clang-format.txt](https://github.com/locusrobotics/fuse/files/14184999/clang-format.txt)

Reply:
All tests are passing now, let me know what you think of the solution, and if I should add some compile warnings/deprecated attributes to anything

Reply:
I've updated that test to be conditional on if manifolds are used or not. It would seem the new Manifold computes the minus jacobian wrt the 1st argument, and LocalParameterization wrt the 2nd so this test. 

The marginal constraint test would change because although LocalParameterization is still supported past 2.1, it is only supported in a manner such that it can be used to generate a manifold.

This leads me to another question for ManifoldAdapter. While converting between the minus conventions is trivial, how can we convert the MinusJacobian? I am not sure/confident that the MinusJacobian is simply negated in every case

Reply:
I think I finally understand, the jacobians shouldnt change between manifold and localparameterization because theyre computing wrt the same variable (they just change spots in the Minus function call). So the only change to the tests are which Minus function is called (which just flips the order of the parameters). So yes the localparameterization test remains the same.

Reply:
I've also updated the local parameterization param naming convention to match the manifold so that its easier to compare across the two